library/pkparse.c

```diff
@@ -1145,12 +1164,17 @@ int mbedtls_pk_parse_key( mbedtls_pk_context *pk,
 {
     int ret;
     const mbedtls_pk_info_t *pk_info;
-
 #if defined(MBEDTLS_PEM_PARSE_C)
     size_t len;
     mbedtls_pem_context pem;
+#endif
 
-    mbedtls_pem_init( &pem );
+    PK_VALIDATE_RET( pk != NULL );
+    PK_VALIDATE_RET( key != NULL || keylen == 0 );
+    PK_VALIDATE_RET( pwd != NULL || pwdlen == 0 );
```

Comment:
Do we test that this function supports `keylen == 0` with a null `key` (same for null `pwd` with `pwdlen == 0`)? It looks like at least most paths through this function, depending on the configuration, will also check `keylen == 0` and reject it. If we already have checks for this elsewhere in the function, why add another check here?

Reply:
You might dismiss this as theoretical babbling, but I think the question of which checks we perform as part of parameter validation should solely depend on the signature and semantics of the respective function, and not on how it's implemented. Also, if you care about code-size, you don't enable this feature in the first place.

Reply:
How its implemented matters because we may break how people expect it to work. We can say "sorry, we didn't promise that in the docs", but we'd still potentially be breaking applications. See https://news.ycombinator.com/item?id=2281932 for a fun story.

Reply:
@Patater I replied to your suggestion of removing checks if they are already performed deeper in the call stack, which I think shouldn't be done.

Reply:
It is tested in the `valid_parameters` test.